### PR TITLE
Revert to exercise start

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
This reverts the files in the editor to the original files of the exercise. In order to do this, we send `GET /api/solutions/:solution_uuid/original_solution_files` to retrieve the files.

### CODE REVIEW

File: app/controllers/test/base_controller.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 class Test::BaseController < ApplicationController
   skip_before_action :verify_authenticity_token
   skip_before_action :authenticate_user!
+
+  disable_site_header!
```
Comment (kntsoriano) [inline app/controllers/test/base_controller.rb:5]: There was a problem with the "Run tests" button not being  seen in the test layout because the header was present. I disabled it for now, and will fix the CSS in another PR.

File: app/javascript/components/Editor.tsx [hunk 2]
```diff
@@ -154,6 +88,11 @@ export function Editor({
 }) {
   const [tab, switchToTab] = useState(TabIndex.INSTRUCTIONS)
   const [theme, setTheme] = useState(Themes.LIGHT)
+  const [
```
Comment (kntsoriano) [inline app/javascript/components/Editor.tsx:91]: I actually think that the reducer for the revert and submission status shouldn't be here and should be abstracted behind a custom hook. I've held off on doing that for now, and will do it in a separate PR.

Comment (kntsoriano) [reply app/javascript/components/Editor.tsx:91]: I also decided to break the state up into two reducers: one that deals with the submission lifecycle, another that deals with the revert lifecycle.

File: app/javascript/components/Editor.tsx [hunk 3]
```diff
@@ -309,14 +251,96 @@ export function Editor({
...
+            revertDispatch({
+              type: RevertStatus.FAILED,
+              payload: { apiError: res.error },
+            })
+          })
+        }
+      })
+  }, [revertDispatch, sendRequest, submission])
+
+  useEffect(() => {
```
Comment (kntsoriano) [inline app/javascript/components/Editor.tsx:317]: When the submission status updates, we set the editor status.

File: app/javascript/components/Editor.tsx [hunk 4]
```diff
@@ -309,14 +251,96 @@ export function Editor({
...
+      case SubmissionStatus.CREATING:
+        setStatus(EditorStatus.CREATING_SUBMISSION)
+        break
+      case SubmissionStatus.CREATING_ITERATION:
+        setStatus(EditorStatus.CREATING_ITERATION)
+        break
+    }
+  }, [submissionStatus])
+
+  useEffect(() => {
```
Comment (kntsoriano) [inline app/javascript/components/Editor.tsx:334]: When the revert status is set, we set the editor status.

File: app/javascript/components/Editor.tsx [hunk 5]
```diff
@@ -309,14 +251,96 @@ export function Editor({
...
+      case RevertStatus.SUCCEEDED:
+        setStatus(EditorStatus.REVERTED)
+        break
+      case RevertStatus.FAILED:
+        setStatus(EditorStatus.REVERT_FAILED)
+        break
+    }
+  }, [revertStatus])
+
+  useEffect(() => {
```
Comment (kntsoriano) [inline app/javascript/components/Editor.tsx:348]: When an error comes in from either reverting and submitting, we set the API error.

Currently, we only have one place in the editor to show API errors. We can decide to have a separate place for each kind of error, but I decided to just put them in one place for now.

File: test/javascript/components/Editor/features/keyboard-shortcuts.test.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+jest.mock('../../../../../app/javascript/components/editor/FileEditor')
```
Comment (kntsoriano) [inline test/javascript/components/Editor/features/keyboard-shortcuts.test.js:1]: I decided to break up the `Editor.test.js` into smaller parts. It was getting too hard just to keep on adding tests to it.

File: app/serializers/serialize_submission.rb [hunk 7]
```diff
@@ -19,6 +19,10 @@ def call
         test_run: Exercism::Routes.api_submission_test_run_url(
           submission.uuid,
           auth_token: user.auth_tokens.first.to_s
+        ),
+        files: Exercism::Routes.api_solution_original_solution_files_url(
```
Comment (iHiD) [inline app/serializers/serialize_submission.rb]:
```suggestion
        files: Exercism::Routes.api_solution_initial_files_url(
```

File: app/controllers/api/solutions/original_solution_files_controller.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+module API
+  module Solutions
+    class OriginalSolutionFilesController < BaseController
```
Comment (iHiD) [inline app/controllers/api/solutions/original_solution_files_controller.rb]:
```suggestion
    class InitialFilesController < BaseController
```

File: app/controllers/api/solutions/original_solution_files_controller.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+      def index
+        solution = current_user.solutions.find_by!(uuid: params[:solution_id])
```
Comment (iHiD) [inline app/controllers/api/solutions/original_solution_files_controller.rb]: See top of api/files_controller for how we do this line elsewhere.

```suggestion
      begin
        solution = Solution.find_by!(uuid: params[:solution_id])
      rescue ActiveRecord::RecordNotFound
        return render_solution_not_found
      end
      
      return render_403(:solution_not_accessible) unless current_user.may_view_solution?(solution)
```

File: app/controllers/test/base_controller.rb (hunk 1 above)
Comment (iHiD) [reply app/controllers/test/base_controller.rb:5]: I think we should just leave the header off anyway. It clashed on the notifications one too. So I'd just leave this as you've done it - not fix the CSS.

**Review by iHiD — APPROVED** [review]
Approving from a Rails POV pending my suggestions :)

No idea on the JS ;)

File: app/controllers/api/solutions/original_solution_files_controller.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+        files = SerializeFiles.(solution.original_solution_files)
+
+        render json: { files: files }
```
Comment (iHiD) [inline app/controllers/api/solutions/original_solution_files_controller.rb]: Please add a controller test that covers the various cases in this. See the other API controller tests for bits you can probably copy/paste in.

File: app/models/solution.rb [hunk 11]
```diff
@@ -59,8 +59,12 @@ def editor_language
     track.slug
   end
 
+  def original_solution_files
```
Comment (iHiD) [inline app/models/solution.rb]:
```suggestion
  def initial_solution_files
```

File: app/models/solution.rb [hunk 12]
```diff
@@ -59,8 +59,12 @@ def editor_language
...
+    git_exercise.editor_solution_files
+  end
+
   def editor_solution_files
-    files = git_exercise.editor_solution_files
+    files = original_solution_files
```
Comment (iHiD) [inline app/models/solution.rb]:
```suggestion
    files = initial_solution_files
```

File: app/models/solution.rb [hunk 13]
```diff
@@ -59,8 +59,12 @@ def editor_language
     track.slug
   end
 
+  def original_solution_files
+    git_exercise.editor_solution_files
+  end
```
Comment (iHiD) [inline app/models/solution.rb:64]: This method needs a test adding.

File: app/serializers/serialize_files.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+  def call
+    files.map do |filename, content|
+      {
+        filename: filename,
+        content: content
+      }
+    end
+  end
+end
```
Comment (iHiD) [inline app/serializers/serialize_files.rb:14]: This needs a test adding.

File: app/serializers/serialize_submission.rb (hunk 7 above)
Comment (iHiD) [inline app/serializers/serialize_submission.rb]:
```suggestion
        initial_files: Exercism::Routes.api_solution_original_solution_files_url(
```

Comment (iHiD) [reply app/serializers/serialize_submission.rb]: Should this not be on the solution, not the submission? This doesn't feel like a property of a submission to me, it feels like a property of the whole solution.

File: app/javascript/components/Editor.tsx (hunk 2 above)
Comment (ErikSchierboom) [reply app/javascript/components/Editor.tsx:91]: Extracting to a custom hook sounds good!

**Review by ErikSchierboom — APPROVED** [review]
This really helps structure things! Separating out code to different files was a very good idea.

File: app/serializers/serialize_submission.rb (hunk 7 above)
Comment (kntsoriano) [reply app/serializers/serialize_submission.rb]: I agree. We pass in the submission to the editor, and not the solution, though. I'll open an issue to look into passing the whole solution instead.
